sklearn/cluster/k_means_.py

```diff
@@ -953,6 +953,13 @@ def fit(self, X, y=None, sample_weight=None):
         """
         random_state = check_random_state(self.random_state)
 
+        # This validates twice but there is not clean way to avoid validation
+        # in k_means(). Please see issue 14897.
```

Comment:
You mean a `_k_means` that gets passed an estimator parameter? Or a `validate` param? the validation is far from trivial because it involves the `copy_x` parameter and later has an influence on the output (`X` is modified inplace if it's true).

Also not the strongest argument, but that file is already cluttered with names. `_k_means.pyx` already exists and is imported as `_k_means`.

In any case, the need for this private helper is an argument in favor of either removing `k_means`, or making it use the estimator.